tests/Unit/Jobs/UpdateAllProductsTest.php

```diff
@@ -0,0 +1,184 @@
+			->willReturn( false );
+
+		$this->merchant_center
+			->method( 'is_connected' )
+			->willReturn( true );
+
+		$this->job->init();
+	}
+
+	public function test_job_name() {
+		$this->assertEquals( self::JOB_NAME, $this->job->get_name() );
+	}
+
+	public function test_single_batched_job_with_items() {
+		$filtered_product_list = new FilteredProductList( $this->generate_simple_products_set( 4 ), 4 );
+
+		$this->action_scheduler->expects( $this->exactly( 2 ) )
+			->method( 'schedule_immediate' )
+			->withConsecutive(
+				array( self::CREATE_BATCH_HOOK, [ 1 ] ),
+				array( self::PROCESS_ITEM_HOOK, [ $filtered_product_list->get_product_ids() ] )
+			);
+
+		$this->product_repository->expects( $this->once() )
+			->method( 'find_sync_ready_products' )
+			->willReturn( $filtered_product_list );
+
+		$this->job->schedule();
+
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+	}
+
+	public function test_single_batched_job_with_no_items() {
+		$filtered_product_list = new FilteredProductList( [], 0 );
+		$this->action_scheduler->expects( $this->once() )
+			->method( 'schedule_immediate' )
+			->with( self::CREATE_BATCH_HOOK, [ 1 ] );
+		$this->product_repository->expects( $this->once() )
+			->method( 'find_sync_ready_products' )
+			->willReturn( $filtered_product_list );
+
+		$this->job->schedule();
+
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+	}
+
+	public function test_multiple_batches_of_items_and_empty_one() {
+
+		/* adding a filter to make batch smaller for testing */
+		add_filter(
+			'woocommerce_gla_batched_job_size',
+			function ( $batch_count, $job_name ) {
+				if ( self::JOB_NAME === $job_name ) {
+					return 2;
+				}
+				return $batch_count;
+			}, 10, 2 );
+
+		$batch_a = new FilteredProductList( $this->generate_simple_products_set( 2 ), 2 );
+		$batch_b = new FilteredProductList( $this->generate_simple_products_set( 2 ), 2 );
+		$batch_c = new FilteredProductList( [], 0 );
+
+		$this->action_scheduler->expects( $this->exactly( 5 ) )
+			->method( 'schedule_immediate' )
+			->withConsecutive(
+				array( self::CREATE_BATCH_HOOK, [ 1 ] ),
+				array( self::PROCESS_ITEM_HOOK, [ $batch_a->get_product_ids() ] ),
+				array( self::CREATE_BATCH_HOOK, [ 2 ] ),
+				array( self::PROCESS_ITEM_HOOK, [ $batch_b->get_product_ids() ] ),
+				array( self::CREATE_BATCH_HOOK, [ 3 ] ),
+			);
+
+		$this->product_repository->expects( $this->exactly( 3 ) )
+			->method( 'find_sync_ready_products' )
+			->withConsecutive( array( [], 2, 0 ), array( [], 2, 2 ), array( [], 2, 4 ) )
+			->willReturnOnConsecutiveCalls( $batch_a, $batch_b, $batch_c );
+
+		$this->job->schedule();
+
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+		do_action( self::CREATE_BATCH_HOOK, 2 );
+		do_action( self::CREATE_BATCH_HOOK, 3 );
+	}
+
+	public function test_process_item() {
+		$filtered_product_list = new FilteredProductList( $this->generate_simple_products_set( 1 ), 1 );
+		$product_id            = $filtered_product_list->get_product_ids()[0];
```

Comment:
This works fine here, but just wanted to mention that we'd need to be careful with this kind of pattern outside of tests. `array_map` preserves the array keys if it's passed only one array, so array key 0 might not always exist especially if some products are filtered out of the list.

Reply:
Got it